osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

```diff
@@ -302,9 +303,41 @@ public ProgressiveCalculationBeatmap(IBeatmap baseBeatmap)
                 this.baseBeatmap = baseBeatmap;
             }
 
-            public readonly List<HitObject> HitObjects = new List<HitObject>();
+            public void AddHitObject(HitObject hitObject)
+            {
+                hitObjects.Add(hitObject);
+
+                var objectType = hitObject.GetType();
+                if (!hitObjectsCounts.ContainsKey(objectType))
+                    hitObjectsCounts[objectType] = 0; // Initialize to 0 if not present
+                hitObjectsCounts[objectType]++;
+            }
+
+            private readonly List<HitObject> hitObjects = new List<HitObject>();
+
+            private readonly Dictionary<Type, int> hitObjectsCounts = new Dictionary<Type, int>();
+
+            public int GetHitObjectCountOf(Type type) => hitObjectsCounts.GetValueOrDefault(type);
+
+            IReadOnlyList<HitObject> IBeatmap.HitObjects => hitObjects;
 
-            IReadOnlyList<HitObject> IBeatmap.HitObjects => HitObjects;
+            private int comboObjectIndex, combo;
+
+            public int GetMaxCombo()
+            {
+                for (; comboObjectIndex < hitObjects.Count; comboObjectIndex++)
+                    addCombo(hitObjects[comboObjectIndex], ref combo);
+                return combo;
+
+                static void addCombo(HitObject hitObject, ref int combo)
+                {
+                    if (hitObject.Judgement.MaxResult.AffectsCombo())
+                        combo++;
+
+                    foreach (var nested in hitObject.NestedHitObjects)
+                        addCombo(nested, ref combo);
+                }
```

Comment:
If you're going to do this you might as well just move the progressive combo accounting to `AddHitObject()` and have this method just return the value of the backing field. Probably simpler that way (gets rid of the `comboObjectIndex` wart).

Reply:
fixed